File: Sources/ExcalidrawMath/Rectangle.swift
import Foundation

/// An axis-aligned rectangle defined by two opposite corners
/// (`packages/math/src/rectangle.ts`).
public struct Rectangle: Equatable, Sendable {
    public var topLeft: Point
    public var bottomRight: Point

    public init(topLeft: Point, bottomRight: Point) {
        self.topLeft = topLeft
        self.bottomRight = bottomRight
    }

    /// Build from min/max bounds (`rectangleFromNumberSequence`).
    public init(minX: Double, minY: Double, maxX: Double, maxY: Double) {
        self.init(topLeft: Point(minX, minY), bottomRight: Point(maxX, maxY))
    }

    /// Intersection points of the rectangle's four edges with a segment
    /// (`rectangleIntersectLineSegment`).
    public func intersection(with segment: LineSegment) -> [Point] {
        let topRight = Point(bottomRight.x, topLeft.y)
        let bottomLeft = Point(topLeft.x, bottomRight.y)
        let edges = [
            LineSegment(topLeft, topRight),
            LineSegment(topRight, bottomRight),
            LineSegment(bottomRight, bottomLeft),
            LineSegment(bottomLeft, topLeft),
        ]
        return edges.compactMap { segment.lineIntersection(with: $0) }
    }

    /// Whether two rectangles overlap (`rectangleIntersectRectangle`).
    public func intersects(_ other: Rectangle) -> Bool {
        topLeft.x < other.bottomRight.x && bottomRight.x > other.topLeft.x &&
            topLeft.y < other.bottomRight.y && bottomRight.y > other.topLeft.y
    }
}

File: web/packages/excalidraw-svelte/src/math/rectangle.ts
import { Point } from "./point.js";
import { LineSegment } from "./segment.js";

/** An axis-aligned rectangle defined by two opposite corners. (parity: Rectangle.swift) */
export class Rectangle {
  constructor(
    public topLeft: Point,
    public bottomRight: Point,
  ) {}

  /** Build from min/max bounds (`rectangleFromNumberSequence`). */
  static fromBounds(minX: number, minY: number, maxX: number, maxY: number): Rectangle {
    return new Rectangle(new Point(minX, minY), new Point(maxX, maxY));
  }

  /** Intersection points of the four edges with a segment (`rectangleIntersectLineSegment`). */
  intersection(segment: LineSegment): Point[] {
    const topRight = new Point(this.bottomRight.x, this.topLeft.y);
    const bottomLeft = new Point(this.topLeft.x, this.bottomRight.y);
    const edges = [
      new LineSegment(this.topLeft, topRight),
      new LineSegment(topRight, this.bottomRight),
      new LineSegment(this.bottomRight, bottomLeft),
      new LineSegment(bottomLeft, this.topLeft),
    ];
    const result: Point[] = [];
    for (const edge of edges) {
      const hit = segment.lineIntersection(edge);
      if (hit !== null) result.push(hit);
    }
    return result;
  }

  /** Whether two rectangles overlap (`rectangleIntersectRectangle`). */
  intersects(other: Rectangle): boolean {
    return (
      this.topLeft.x < other.bottomRight.x &&
      this.bottomRight.x > other.topLeft.x &&
      this.topLeft.y < other.bottomRight.y &&
      this.bottomRight.y > other.topLeft.y
    );
  }
}
